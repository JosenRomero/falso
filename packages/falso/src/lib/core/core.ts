import { random } from '../random';

export interface FakeOptions {
  length?: number;
  priority?: 'length' | 'unique';
}

type Return<T, O extends FakeOptions> = [O] extends [never]
  ? T
  : O['length'] extends number
  ? T[]
  : T;

export function fake<T, Options extends FakeOptions>(
  data: T[] | (() => T),
  options?: Options,
  comparisonFunction: (item: T, items: T[]) => boolean = primitiveValueIsUnique,
  comparisonKeys?: string[]
): Return<T, Options> {
  if (Array.isArray(data)) {
    return fakeFromArray(data, options) as any;
  }

  return fakeFromFunction(
    data,
    comparisonFunction,
    comparisonKeys,
    options
  ) as any;
}

export function fakeFromFunction<T, Options extends FakeOptions>(
  data: () => T,
  isItemADuplicateFunction: (
    item: T,
    items: T[],
    comparisonKeys?: string[]
  ) => boolean,
  comparisonKeys?: string[],
  options?: Options
) {
  if (!options?.length) {
    return data();
  }

  const priority = options?.priority ?? 'length';

  if (priority === 'length') {
    return Array.from({ length: options.length }, (_, index) => data());
  }

  const items: T[] = [];

  let attempts = 0;
  const maxAttempts = options.length * 2;

  while (items.length < options.length && attempts < maxAttempts) {
    const item = data();

    if (!isItemADuplicateFunction(item, items, comparisonKeys)) {
      items.push(item);
    }

    attempts++;
  }

  return items;
}

export function fakeFromArray<T, Options extends FakeOptions>(
  data: T[],
  options?: Options
) {
  if (!options?.length) {
    return randElement(data);
  }

  const priority = options?.priority ?? 'length';

  if (priority === 'length') {
    return Array.from({ length: options.length }, () => randElement(data));
  }

  const clonedData: T[] = JSON.parse(JSON.stringify(data));
  const newArray: T[] = [];

  while (clonedData.length && newArray.length !== options.length) {
    const randomIndex = Math.floor(random() * clonedData.length);
    const item = clonedData[randomIndex];

    newArray.push(item);
    clonedData.splice(randomIndex, 1);
  }

  return newArray;
}

export const primitiveValueIsUnique: <T>(item: T, items: T[]) => boolean = (
  item,
  items
) => !items.includes(item);

export const dateIsUnique: (date: Date, dates: Date[]) => boolean = (
  date,
  dates
) => !dates.some((d) => d.valueOf() === date.valueOf());

export const checkUniqueObjectWithId: <T extends { id: string }>(
  item: T,
  items: T[]
) => boolean = (item, items) => objectIsUnique(item, items, ['id']);

export const objectIsUnique: (
  item: any,
  items: any[],
  keys: string[]
) => boolean = (item: any, items: any[], keys: string[]) => {
  for (const key of keys) {
    if (!item[key]) {
      throw new Error(`${key} does not exist in this array value type`);
    }

    if (items.some((arrayItem) => arrayItem[key] === item[key])) {
      return true;
    }
  }

  return false;
};

export function randElement<T>(arr: T[]): T {
  return arr[Math.floor(random() * arr.length)];
}

export interface RandomInRangeOptions {
  min?: number;
  max?: number;
  fraction?: number;
}

export function getRandomInRange({
  min = 1.0,
  max = 9999.99,
  fraction = 0,
}: RandomInRangeOptions = {}) {
  return Number((random() * (max - min) + min).toFixed(fraction));
}
