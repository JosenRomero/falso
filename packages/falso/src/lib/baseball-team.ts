import { FakeOptions, fake } from './core/core';
import { data } from './baseball-team.json';

/**
 * Generate a random baseball team.
 *
 * @category sports
 *
 * @example
 *
 * randBaseballTeam()
 *
 * @example
 *
 * randBaseballTeam({ length: 10 })
 *
 * @example
 *
 * randBaseballTeam({ length: 10, priority: 'unique' }) // default is 'length' ('length' | 'unique')
 *
 */
export function randBaseballTeam<Options extends FakeOptions = never>(
  options?: Options
) {
  return fake(data, options);
}
