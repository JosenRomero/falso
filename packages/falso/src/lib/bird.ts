import { FakeOptions, fake } from './core/core';
import { data } from './bird.i18n.json';

/**
 * Generate a random bird.
 *
 * @category animals
 *
 * @example
 *
 * randBird()
 *
 * @example
 *
 * randBird({ length: 10 })
 *
 * @example
 *
 * randBird({ length: 10, priority: 'unique' }) // default is 'length' ('length' | 'unique')
 *
 */
export function randBird<Options extends FakeOptions = never>(
  options?: Options
) {
  return fake(data, options);
}
