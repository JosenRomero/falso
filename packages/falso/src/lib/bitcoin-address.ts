import { fake, FakeOptions } from './core/core';
import { randSequence } from './sequence';

/**
 * Generate a random bitcoin-address.
 *
 * @category finance
 *
 * @example
 *
 * randBitcoinAddress()
 *
 * @example
 *
 * randBitcoinAddress({ length: 10 })
 *
 * @example
 *
 * randBitcoinAddress({ length: 10, priority: 'unique' }) // default is 'length' ('length' | 'unique')
 *
 */
export function randBitcoinAddress<Options extends FakeOptions = never>(
  options?: Options
) {
  return fake(() => randSequence({ size: 33 }), options);
}
