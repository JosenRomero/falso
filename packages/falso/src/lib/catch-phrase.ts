import { FakeOptions, fake } from './core/core';
import { data } from './catch-phrase.json';

/**
 * Generate a random catch phrase.
 *
 * @category text
 *
 * @example
 *
 * randCatchPhrase()
 *
 * @example
 *
 * randCatchPhrase({ length: 10 })
 *
 * @example
 *
 * randCatchPhrase({ length: 10, priority: 'unique' }) // default is 'length'
 *
 */
export function randCatchPhrase<Options extends FakeOptions = never>(
  options?: Options
) {
  return fake(data, options);
}
