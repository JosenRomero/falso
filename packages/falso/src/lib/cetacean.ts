import { FakeOptions, fake } from './core/core';
import { data } from './cetacean.json';

/**
 * Generate a random cetacean.
 *
 * @category animals
 *
 * @example
 *
 * randCetacean()
 *
 * @example
 *
 * randCetacean({ length: 10 })
 *
 * @example
 *
 * randCetacean({ length: 10, priority: 'unique' }) // default is 'length' ('length' | 'unique')
 *
 */
export function randCetacean<Options extends FakeOptions = never>(
  options?: Options
) {
  return fake(data, options);
}
