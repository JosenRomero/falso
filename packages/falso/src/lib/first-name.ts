import { fake } from './core/core';
import { data } from './first-name.json';
import { NameOptions } from './full-name';
import { randBoolean } from './boolean';

/**
 * Generate a random first name.
 *
 * @category person
 *
 * @example
 *
 * randFirstName()
 *
 * @example
 *
 * randFirstName({ withAccents: true })
 *
 * @example
 *
 * randFirstName({ length: 10 })
 *
 * @example
 *
 * randFirstName({ length: 10, priority: 'unique' }) // default is 'length' ('length' | 'unique')
 *
 */
export function randFirstName<Options extends NameOptions = never>(
  options?: Options
) {
  const withAccents = options?.withAccents ?? randBoolean();
  const names = withAccents ? data['withAccents'] : data['withoutAccents'];

  return fake(names, options);
}
