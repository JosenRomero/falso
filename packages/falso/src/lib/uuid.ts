import { v4 as uuidv4 } from 'uuid';
import { fake, FakeOptions } from './core/core';

/**
 * Generate a random uuid.
 *
 * @category general, database
 *
 * @example
 *
 * randUuid()
 *
 * @example
 *
 * randUuid({ length: 10 })
 *
 * @example
 *
 * randUuid({ length: 10, priority: 'unique' }) // default is 'length' ('length' | 'unique')
 *
 */
export function randUuid<Options extends FakeOptions = never>(
  options?: Options
) {
  return fake(() => uuidv4(), options);
}
