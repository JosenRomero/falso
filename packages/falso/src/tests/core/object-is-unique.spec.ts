import { objectIsUnique } from '../../lib/core/core';
import { randUser, User } from '../../lib/user';
import { randUuid } from '../../lib/uuid';
import { randFirstName } from '../../lib/first-name';

describe('objectIsUnique', () => {
  describe("keys contains a key that doesn't exist", () => {
    let array: User[];
    let newItem: User;
    let keys: string[];

    beforeEach(() => {
      array = randUser({ length: 3 });
      newItem = randUser();
      keys = ['id', 'noExistentKey', 'firstName'];
    });

    it('should throw error', () => {
      expect(() =>
        objectIsUnique(newItem, array, keys)
      ).toThrowErrorMatchingSnapshot();
    });
  });

  describe('1 key is passed', () => {
    let array: User[];
    let newItem: User;
    let keys: string[];

    beforeEach(() => {
      array = randUser({ length: 3 });
      newItem = randUser();
      keys = ['id'];
    });

    describe('passed item matches and array item have 1 matching key value', () => {
      let sharedId: string;

      beforeEach(() => {
        sharedId = randUuid();
        newItem.id = sharedId;
        array[1].id = sharedId;
      });

      it('should return true', () => {
        const result = objectIsUnique(newItem, array, keys);

        expect(result).toEqual(true);
      });
    });

    describe('passed item matches and array item have 0 matching key values', () => {
      beforeEach(() => {
        array[0].id = randUuid() + '0';
        array[1].id = randUuid() + '1';
        array[2].id = randUuid() + '2';
        newItem.id = randUuid() + '3';
      });

      it('should return true', () => {
        const result = objectIsUnique(newItem, array, keys);

        expect(result).toEqual(false);
      });
    });
  });

  describe('multiple keys are passed', () => {
    let array: User[];
    let newItem: User;
    let keys: string[];

    beforeEach(() => {
      array = randUser({ length: 3 });
      newItem = randUser();
      keys = ['id', 'firstName'];
    });

    describe('passed item matches and array item have 1 matching key value', () => {
      let sharedFirstName: string;

      beforeEach(() => {
        sharedFirstName = randFirstName();
        newItem.firstName = sharedFirstName;
        array[1].firstName = sharedFirstName;
      });

      it('should return true', () => {
        const result = objectIsUnique(newItem, array, keys);

        expect(result).toEqual(true);
      });
    });

    describe('passed item matches and array item have 0 matching key values', () => {
      beforeEach(() => {
        array[0].id = randUuid() + '0';
        array[0].firstName = randFirstName() + '0';
        array[1].id = randUuid() + '1';
        array[1].id = randFirstName() + '1';
        array[2].id = randUuid() + '2';
        array[2].id = randFirstName() + '2';
        newItem.id = randUuid() + '3';
        newItem.id = randFirstName() + '3';
      });

      it('should return true', () => {
        const result = objectIsUnique(newItem, array, keys);

        expect(result).toEqual(false);
      });
    });
  });
});
