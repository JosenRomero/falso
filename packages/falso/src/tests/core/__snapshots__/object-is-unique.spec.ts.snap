// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`objectIsUnique keys contains a key that doesn't exist should throw error 1`] = `"noExistentKey does not exist in this array value type"`;
